cpp/src/arrow/extension/fixed_shape_tensor.cc

```diff
@@ -86,7 +88,7 @@ bool FixedShapeTensorType::ExtensionEquals(const ExtensionType& other) const {
   if (extension_name() != other.extension_name()) {
     return false;
   }
-  const auto& other_ext = static_cast<const FixedShapeTensorType&>(other);
+  const auto& other_ext = dynamic_cast<const FixedShapeTensorType&>(other);
```

Comment:
Did it fail otherwise? I'm a bit curious to know if the change to `dynamic_cast` was necessary.

(ideally this should be a `checked_cast`: i.e. dynamically checked in debug mode, not in release mode)

Reply:
I forget what the reason was, I vaguely  remember it had to do with a change I later reverted.
Changed to `checked_cast` and it seems to work fine!